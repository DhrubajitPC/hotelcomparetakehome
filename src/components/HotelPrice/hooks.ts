import { useComparisons } from "../../hooks";
import { useAppContext } from "../../state/Context";

export const usePriceBreakdown = (hotelId: number) => {
  const {
    state: { prices },
  } = useAppContext();
  const price = prices.find((p) => p.id === hotelId);
  return {
    totalPrice: price?.price ?? "Rates Unavailable",
    tax: price?.taxes_and_fees?.tax,
    hotelFees: price?.taxes_and_fees?.hotel_fees,
  };
};

export const useSavings = (hotelId: number) => {
  const comparisons = useComparisons(hotelId);

  if (comparisons.length === 0) return null;

  if (comparisons[comparisons.length - 1].name === "Us") return null;

  const ourPrice = comparisons.find((c) => c.name === "Us")?.price;
  if (!ourPrice) return null;

  const maxPrice = comparisons[comparisons.length - 1].price;

  return ((maxPrice - ourPrice) / maxPrice) * 100;
};
