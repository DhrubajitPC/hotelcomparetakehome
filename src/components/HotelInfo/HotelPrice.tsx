import { FC } from "react";
import { useAppContext } from "../../state/Context";
import TaxBreakdown from "../TaxBreakdown";
import Tooltip from "../Tooltip";
import { usePriceBreakdown } from "./hooks";
import { formatPrice } from "./utils";

const HotelPrice: FC<{ hotelId: number }> = ({ hotelId }) => {
  const { state } = useAppContext();
  const currency = state.selectedCurrency;
  const priceBreakdown = usePriceBreakdown(hotelId);

  const isTaxInclusive = Boolean(priceBreakdown.hotelFees);
  const showCurrency = priceBreakdown.totalPrice !== "Rates Unavailable";

  let price;
  if (showCurrency) {
    price = formatPrice(currency, priceBreakdown.totalPrice as number);
  } else {
    price = priceBreakdown.totalPrice;
  }

  return isTaxInclusive ? (
    <Tooltip
      body={
        <TaxBreakdown
          tax={priceBreakdown.tax!}
          hotelFees={priceBreakdown.hotelFees!}
        />
      }
    >
      <p style={{ textDecoration: "underline" }}>
        {showCurrency && currency} {price}*
      </p>
    </Tooltip>
  ) : (
    <p>
      {showCurrency && currency} {price}
    </p>
  );
};

export default HotelPrice;
