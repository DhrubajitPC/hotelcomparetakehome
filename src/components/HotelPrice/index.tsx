import { FC } from "react";
import { useAppContext } from "../../state/Context";
import { formatPrice } from "../../utils";
import TaxBreakdown from "../TaxBreakdown";
import Tooltip from "../Tooltip";
import { usePriceBreakdown, useSavings } from "./hooks";

const HotelPrice: FC<{ hotelId: number }> = ({ hotelId }) => {
  const {
    state: { selectedCurrency: currency },
  } = useAppContext();
  const priceBreakdown = usePriceBreakdown(hotelId);
  const savings = useSavings(hotelId);

  const isTaxInclusive = Boolean(priceBreakdown.hotelFees);
  const showCurrency = priceBreakdown.totalPrice !== "Rates Unavailable";

  let price;
  if (showCurrency) {
    price = formatPrice(currency, priceBreakdown.totalPrice as number);
  } else {
    price = priceBreakdown.totalPrice;
  }

  return (
    <>
      {isTaxInclusive ? (
        <Tooltip
          body={
            <TaxBreakdown
              tax={priceBreakdown.tax!}
              hotelFees={priceBreakdown.hotelFees!}
            />
          }
        >
          <p style={{ textDecoration: "underline", color: "#eb4d4b" }}>
            {showCurrency && currency} {price}*
          </p>
        </Tooltip>
      ) : (
        <p style={{ color: "#eb4d4b" }}>
          Price: {showCurrency && currency} {price}
        </p>
      )}

      {savings && (
        <p style={{ textTransform: "uppercase", fontSize: "14px" }}>
          you save {savings.toFixed(2)}%
        </p>
      )}
      <button
        style={{
          padding: "20px",
          background: "#4834d4",
          border: "none",
          borderRadius: "10px",
          fontSize: "18px",
          color: "#fff",
        }}
        onClick={() => {}}
      >
        Book!
      </button>
    </>
  );
};

export default HotelPrice;
