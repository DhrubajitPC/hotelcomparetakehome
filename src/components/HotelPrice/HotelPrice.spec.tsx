import { fireEvent, render, screen } from "@testing-library/react";
import HotelPrice from ".";

let mockPrice: Array<Record<any, any>> = [
  {
    id: 2,
    price: 164,
    competitors: {
      Traveloka: 190,
      Expedia: 163,
    },
    taxes_and_fees: {
      tax: 13.12,
      hotel_fees: 16.4,
    },
  },
];

let mockPriceBreakdown: Record<any, any> = {
  totalPrice: 120,
  tax: 11.11,
  hotelFees: 22,
};

jest
  .mock("./hooks", () => ({
    __esModule: true,
    usePriceBreakdown: () => mockPriceBreakdown,
    useSavings: () => 10,
  }))
  .mock("../../state/Context", () => ({
    __esModule: true,
    useAppContext: () => ({
      state: {
        prices: mockPrice,
        selectedCurrency: "SGD",
      },
    }),
  }));

describe("#HotelPrice", () => {
  it("should render hotelPrice component correctly when there is a breakdown", () => {
    const { container } = render(<HotelPrice hotelId={2} />);
    const tooltipWrapper = screen.getByTestId("tooltip-container");
    fireEvent.mouseEnter(tooltipWrapper);
    expect(container).toMatchSnapshot();
  });

  it("should render hotelPrice component correctly when there is no price", () => {
    mockPriceBreakdown = {
      totalPrice: "Rates Unavailable",
      tax: undefined,
      hotelFees: undefined,
    };

    const { container } = render(<HotelPrice hotelId={2} />);
    expect(container).toMatchSnapshot();
  });

  it("should render hotelPrice component correctly when there is no breakdown", () => {
    mockPriceBreakdown = {
      totalPrice: 120,
      tax: undefined,
      hotelFees: undefined,
    };

    const { container } = render(<HotelPrice hotelId={2} />);
    expect(container).toMatchSnapshot();
  });
});
