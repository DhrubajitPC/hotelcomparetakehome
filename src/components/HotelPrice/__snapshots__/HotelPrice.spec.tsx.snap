// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#HotelPrice should render hotelPrice component correctly when there is a breakdown 1`] = `
<div>
  <div
    data-testid="tooltip-container"
    style="position: relative; cursor: help;"
  >
    <p
      style="text-decoration: underline; color: rgb(235, 77, 75);"
    >
      SGD
       
      120
      *
    </p>
    <div
      class="body"
    >
      <p>
        Tax: 
        11.11
      </p>
      <p>
        Hotel Fees: 
        22
      </p>
    </div>
  </div>
  <p
    style="text-transform: uppercase; font-size: 14px;"
  >
    you save 
    10.00
    %
  </p>
  <button
    style="padding: 20px; background: rgb(72, 52, 212); border-radius: 10px; font-size: 18px; color: rgb(255, 255, 255); cursor: pointer;"
  >
    Book!
  </button>
</div>
`;

exports[`#HotelPrice should render hotelPrice component correctly when there is no breakdown 1`] = `
<div>
  <p
    style="color: rgb(235, 77, 75);"
  >
    Price: 
    SGD
     
    120
  </p>
  <p
    style="text-transform: uppercase; font-size: 14px;"
  >
    you save 
    10.00
    %
  </p>
  <button
    style="padding: 20px; background: rgb(72, 52, 212); border-radius: 10px; font-size: 18px; color: rgb(255, 255, 255); cursor: pointer;"
  >
    Book!
  </button>
</div>
`;

exports[`#HotelPrice should render hotelPrice component correctly when there is no price 1`] = `
<div>
  <p
    style="color: rgb(235, 77, 75);"
  >
    Price: 
     
    Rates Unavailable
  </p>
  <p
    style="text-transform: uppercase; font-size: 14px;"
  >
    you save 
    10.00
    %
  </p>
  <button
    style="padding: 20px; background: rgb(72, 52, 212); border-radius: 10px; font-size: 18px; color: rgb(255, 255, 255); cursor: pointer;"
  >
    Book!
  </button>
</div>
`;
